### PR TITLE
Primary caching 18: range invalidation (ENABLED BY DEFAULT :confetti_ball:)

### LINKED ISSUE #374: Query caching for "sticky" scenes (plots, text logs...)
Currently, querying a sticky scene is O(n), and as time goes on it gets more and more laggy.
This is already an issue for plots and text logs today.

Ideally, one would want to cache the results of a query up to `now()`, and then only query the difference the next frame, and so on and so on.
Problem is: what about out of order data?

### LINKED ISSUE #4809: Primary cache: range invalidation
Title.

### PR DESCRIPTION
**Pull request by teh-cmc**
Implement range invalidation and do a quality pass over all the size tracking stuff in the cache.

**Range caching is now enabled by default!**

- Fixes #4809 
- Fixes #374

---

Part of the primary caching series of PR (index search, joins, deserialization):
- #4592
- #4593
- #4659
- #4680 
- #4681
- #4698
- #4711
- #4712
- #4721 
- #4726 
- #4773
- #4784
- #4785
- #4793
- #4800
- #4851
- #4852
- #4853
- #4856

---

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested the web demo (if applicable):
  * Using newly built examples: [app.rerun.io](https://app.rerun.io/pr/4800/index.html)
  * Using examples from latest `main` build: [app.rerun.io](https://app.rerun.io/pr/4800/index.html?manifest_url=https://app.rerun.io/version/main/examples_manifest.json)
  * Using full set of examples from `nightly` build: [app.rerun.io](https://app.rerun.io/pr/4800/index.html?manifest_url=https://app.rerun.io/version/nightly/examples_manifest.json)
* [x] The PR title and labels are set such as to maximize their usefulness for the next release's CHANGELOG

- [PR Build Summary](https://build.rerun.io/pr/4800)
- [Docs preview](https://rerun.io/preview/45f8d624e6867d68980eb747612484936ea0e127/docs) 
- [Examples preview](https://rerun.io/preview/45f8d624e6867d68980eb747612484936ea0e127/examples) 
- [Recent benchmark results](https://build.rerun.io/graphs/crates.html)
- [Wasm size tracking](https://build.rerun.io/graphs/sizes.html)

### CODE REVIEW

File: crates/re_query_cache/src/cache.rs [hunk 1]
```diff
@@ -129,7 +129,17 @@ impl Caches {
                 let mut caches = caches.0.write();
 
                 let caches_per_archetype = caches.entry(key.clone()).or_default();
-                caches_per_archetype.handle_pending_invalidation(&key);
+
+                // TODO: we cannot be holding `caches` when doing this.
```
Comment (teh-cmc) [inline crates/re_query_cache/src/cache.rs:133]: Remove: locks have been cleaned up in next PR.

File: crates/re_query_cache/src/cache.rs [hunk 2]
```diff
@@ -166,7 +176,17 @@ impl Caches {
...
```
Comment (teh-cmc) [inline crates/re_query_cache/src/cache.rs:180]: Remove: locks have been cleaned up in next PR.

File: crates/re_query_cache/src/flat_vec_deque.rs [hunk 3]
```diff
@@ -45,9 +45,15 @@ pub trait ErasedFlatVecDeque: std::any::Any {
     /// This is prefixed with `dyn_` to avoid method dispatch ambiguities that are very hard to
     /// avoid even with explicit syntax and that silently lead to infinite recursions.
     fn dyn_truncate(&mut self, at: usize);
+
+    /// Dynamically dispatches to [`<FlatVecDeque<T> as SizeBytes>::total_size_bytes(self)`].
```
Comment (teh-cmc) [inline crates/re_query_cache/src/flat_vec_deque.rs:49]: That's a valid link? oO anyway, remove the `(self)` part

File: crates/re_query_cache/tests/range.rs
Comment (teh-cmc) [inline crates/re_query_cache/tests/range.rs:1]: Exact same test suite as in `tests/latest_at.rs` but with range queries instead.

File: crates/re_viewer_context/src/app_options.rs [hunk 4]
```diff
@@ -65,7 +65,7 @@ impl Default for AppOptions {
             experimental_additive_workflow: cfg!(debug_assertions),
 
             experimental_primary_caching_latest_at: true,
-            experimental_primary_caching_range: false,
+            experimental_primary_caching_range: true,
```
Comment (teh-cmc) [inline crates/re_viewer_context/src/app_options.rs:68]: :partying_face:

**Review by jleibs — APPROVED** [review]

**Pull request merged**